loadflow/loadflow-validation/src/main/java/com/powsybl/loadflow/validation/BalanceTypeGuesser.java

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.loadflow.validation;
+
+import static java.util.Comparator.comparingDouble;
+
+import java.util.AbstractMap.SimpleEntry;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+import com.powsybl.iidm.network.Generator;
+import com.powsybl.iidm.network.Network;
+
+/**
+ *
+ * @author Massimo Ferraro <massimo.ferraro@techrain.eu>
+ */
+public class BalanceTypeGuesser {
+
+    private static final String NON_EXISTENT_SLACK_ID = "NonExistentSlackBusId";
+
+    private String maxChangeId = "";
+    private double maxChange = 0;
+    private double sumMaxP = 0;
+    private double sumTargetP = 0;
+    private double sumP = 0;
+    private double cumSumVarKMax = 0;
+    private double cumSumVarKTarget = 0;
+    private double cumSumVarKHeadroom = 0;
+    private int sumChanged = 0;
+
+    private BalanceType balanceType;
+    private String slack;
+
+    private double kMax = 0;
+    private double kTarget = 0;
+    private double kHeadroom = 0;
+
+    public BalanceTypeGuesser() {
+        this.balanceType = BalanceType.NONE;
+        this.slack = NON_EXISTENT_SLACK_ID;
+    }
+
+    public BalanceTypeGuesser(Network network, double threshold) {
+        Objects.requireNonNull(network);
+        guess(network, threshold);
+    }
+
+    private void guess(Network network, double threshold) {
+        network.getGeneratorStream().forEach(generator -> computeSums(generator, threshold));
+        if (sumChanged > 0) {
+            kMax = (sumP - sumTargetP) / sumMaxP;
+            kTarget = (sumP - sumTargetP) / sumTargetP;
+            kHeadroom = (sumP - sumTargetP) / (sumMaxP - sumTargetP);
+            double varKMax = (cumSumVarKMax / sumMaxP) / (Math.max(1E-12, kMax * kMax) - 1);
+            double varKTarget = (cumSumVarKTarget / sumTargetP) / (Math.max(1E-12, kTarget * kTarget) - 1);
+            double varKHeadroom = (cumSumVarKHeadroom / (sumMaxP - sumTargetP)) / (Math.max(1E-12, kHeadroom * kHeadroom) - 1);
```

Comment:
The logic to estimate the relevance of the balancing mode seems the same for each of the 3 candidates (is it chi 2?).
Shouldn't we try to mutualize this code instead of repeating 3 times the logic ?

Reply:
It is the variance of k divided by the square of the expectancy of k. Without the -1, it is indeed a chi². I reinvented the wheel...